lib/tbot/workloadidentity/workloadattest/sigstore.go

```diff
@@ -0,0 +1,234 @@
+	keychain, err := sigstore.Keychain(cfg.CredentialsPath)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading credentials")
+	}
+
+	regHosts := make([]string, len(cfg.AdditionalRegistries))
+	for idx, reg := range cfg.AdditionalRegistries {
+		regHosts[idx] = reg.Host
+	}
+
+	att := &SigstoreAttestor{
+		cfg:                cfg,
+		log:                log,
+		registryHosts:      regHosts,
+		keychain:           keychain,
+		maxRefreshInterval: sigstoreMaxRefreshInterval,
+		failures:           make(map[string]time.Time),
+	}
+	att.cache, err = lru.NewWithEvict[string, *workloadidentityv1.WorkloadAttrsSigstore](64, att.onCacheEviction)
+	if err != nil {
+		return nil, trace.Wrap(err, "building LRU cache")
+	}
+	return att, nil
+
+}
+
+// Attest discovers signatures and attestations for the given container.
+func (a *SigstoreAttestor) Attest(ctx context.Context, ctr Container) (*workloadidentityv1.WorkloadAttrsSigstore, error) {
+	logger := a.log.With(
+		"image", ctr.GetImage(),
+		"image_digest", ctr.GetImageDigest(),
+	)
+	logger.InfoContext(ctx, "Starting Sigstore workload attestation")
+
+	cached, hit := a.getCached(ctr.GetImageDigest())
+	if hit {
+		logger.DebugContext(ctx, "Sigstore attestor cache hit")
+		return cached, nil
+	}
+
+	logger.DebugContext(ctx, "Sigstore attestor cache miss")
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	payloads, err := sigstore.Discover(ctx,
+		ctr.GetImage(),
+		ctr.GetImageDigest(),
+		sigstore.DiscoveryConfig{
+			Logger:               a.log,
+			Keychain:             a.keychain,
+			AdditionalRegistries: a.registryHosts,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err, "discovering Sigstore payloads")
+	}
+
+	result := &workloadidentityv1.WorkloadAttrsSigstore{
+		Payloads: payloads,
+	}
+	_ = a.cache.Add(ctr.GetImageDigest(), result)
+
+	a.failuresMu.Lock()
+	delete(a.failures, ctr.GetImageDigest())
+	a.failuresMu.Unlock()
+
+	return result, nil
+}
+
+// MarkFailed tracks that we failed to get a workload identity for the given
+// container. After sigstoreMaxRefreshInterval has elapsed, we will attempt to
+// refresh the signatures and bundles.
+func (a *SigstoreAttestor) MarkFailed(ctx context.Context, ctr Container) {
+	a.log.DebugContext(ctx,
+		"Marking image digest as failed in Sigstore attestor cache",
+		"image", ctr.GetImage(),
+		"image_digest", ctr.GetImageDigest(),
+	)
+	a.failuresMu.Lock()
+	defer a.failuresMu.Unlock()
+
+	if _, ok := a.failures[ctr.GetImageDigest()]; !ok {
+		a.failures[ctr.GetImageDigest()] = time.Now()
+	}
+}
+
+func (a *SigstoreAttestor) getCached(imageDigest string) (*workloadidentityv1.WorkloadAttrsSigstore, bool) {
+	cached, ok := a.cache.Get(imageDigest)
+	if !ok {
+		return nil, false
+	}
+
+	a.failuresMu.Lock()
+	defer a.failuresMu.Unlock()
+
+	failedAt, failed := a.failures[imageDigest]
+	if !failed || time.Since(failedAt) < a.maxRefreshInterval {
+		return cached, true
+	}
+	return nil, false
+}
```

Comment:
More of a nit so feel free to disregard, but does the bool return provide any value here vs just returning a nil?

Reply:
Yeah that's a fair point! Will remove the second return value.